src/hooks/sync-sites/sync-sites-context.tsx

```diff
@@ -1,24 +1,67 @@
-import React, { createContext, useContext, useState } from 'react';
+import { __, sprintf } from '@wordpress/i18n';
+import React, { createContext, useCallback, useContext, useState } from 'react';
+import { getIpcApi } from '../../lib/get-ipc-api';
 import { SyncSite } from '../use-fetch-wpcom-sites';
-import { usePullPushTimestamps } from '../use-pull-push-timestamps';
+import { useFormatLocalizedTimestamps } from '../use-format-localized-timestamps';
 import { useListenDeepLinkConnection } from './use-listen-deep-link-connection';
 import { useSiteSyncManagement } from './use-site-sync-management';
 import { useSyncPull } from './use-sync-pull';
 import { useSyncPush } from './use-sync-push';
 
 export type SyncSitesContextType = ReturnType< typeof useSyncPull > &
 	ReturnType< typeof useSyncPush > &
-	ReturnType< typeof useSiteSyncManagement > &
-	ReturnType< typeof usePullPushTimestamps >;
+	ReturnType< typeof useSiteSyncManagement > & {
+		getLastSyncTimeText: ( timestamp: string | null, type: 'pull' | 'push' ) => string;
+		updateSiteTimestamp: ( site: SyncSite, type: 'pull' | 'push' ) => Promise< void >;
+	};
 
 const SyncSitesContext = createContext< SyncSitesContextType | undefined >( undefined );
 
 export function SyncSitesProvider( { children }: { children: React.ReactNode } ) {
+	const { formatRelativeTime } = useFormatLocalizedTimestamps();
 	const [ pullStates, setPullStates ] = useState< SyncSitesContextType[ 'pullStates' ] >( {} );
+	const [ connectedSites, setConnectedSites ] = useState< SyncSite[] >( [] );
+
+	const getLastSyncTimeText = useCallback(
+		( timestamp: string | null, type: 'pull' | 'push' ): string => {
+			if ( timestamp === null ) {
+				return type === 'pull'
+					? __( 'You have not pulled this site yet.' )
+					: __( 'You have not pushed this site yet.' );
+			}
+
+			return sprintf(
+				type === 'pull'
+					? __( 'You pulled this site %s ago.' )
+					: __( 'You pushed this site %s ago.' ),
+				formatRelativeTime( timestamp )
+			);
+		},
+		[ formatRelativeTime ]
+	);
+
+	const updateSiteTimestamp = useCallback( async ( site: SyncSite, type: 'pull' | 'push' ) => {
+		try {
+			const updatedSite = {
+				...site,
+				[ type === 'pull' ? 'lastPullTimestamp' : 'lastPushTimestamp' ]: new Date().toISOString(),
+			};
+
+			await getIpcApi().updateSingleConnectedWpcomSite( updatedSite );
+			setConnectedSites( ( sites ) =>
+				sites.map( ( s ) => ( s.id === site.id ? updatedSite : s ) )
+			);
+		} catch ( error ) {
+			console.error( 'Failed to update timestamp:', error );
+		}
+	}, [] );
+
 	const { pullSite, isAnySitePulling, isSiteIdPulling, clearPullState, getPullState } = useSyncPull(
 		{
 			pullStates,
 			setPullStates,
+			onPullSuccess: ( site ) => updateSiteTimestamp( site, 'pull' ),
```

Comment:
```suggestion
			onPullSuccess: ( siteId ) =>
				updateSiteTimestamp(
					connectedSites.find( ( { id } ) => siteId === id ),
					'pull'
				),
```
This doesn't work out-of-the-box, because TS complains that `Array.prototype.find` could return undefined. Still, it simplifies the solution as a whole, because we don't need to pass `connectedSite` around inside `use-sync-pull.ts`.

We could solve the TS error by making `updateSiteTimestamp` return early if `site` is undefined.

Reply:
It is something that I am testing now, it might work but we would need to ensure that we also pass selectedSiteId as a site with the same remoteSiteId can be attached to a couple of different local sites.

Reply:
I know I left a few different comments on this topic, but I would start with [this one](https://github.com/Automattic/studio/pull/714#discussion_r1871021613), since I think the concept is pretty clear.

> ensure that we also pass selectedSiteId

Good point. Assuming we get that "for free" with the current solution (because of how the closures are structured), I think it would be an additional benefit if we could make that dependency clearer.

Reply:
Implemented it in https://github.com/Automattic/studio/pull/714/commits/40724600c3658b59c8a39e6b72aba8777c271246 since we have localSiteId available there already. Thanks for the suggestion 👍